app/controllers/concerns/shopify_app/domain_protection.rb

```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+module ShopifyApp
+  module DomainProtection
+    extend ActiveSupport::Concern
+
+    included do
+      before_action :check_shop_domain
+      before_action :check_shop_known
+    end
+
+    def current_shopify_domain
+      return if params[:shop].blank?
+      @shopify_domain ||= ShopifyApp::Utils.sanitize_shop_domain(params[:shop])
+    end
+
+    private
+
+    def check_shop_domain
+      redirect_to(ShopifyApp.configuration.login_url) unless current_shopify_domain
+    end
+
+    def check_shop_known
+      @shop = Shop.find_by(shopify_domain: @shopify_domain)
+      redirect_to("#{ShopifyApp.configuration.login_url}?shop=#{@shopify_domain}") unless @shop
```

Comment:
Should we add a `return_to` parameter so that the redirect returns us back to the page it was last at and not always the homepage? Email folks added a PR fix for this yesterday: https://github.com/Shopify/email/pull/4746

Reply:
@rezaansyed I think if the shop doesn't exist we'd want them to go through the login flow again, right?
Email does that because it is trying to masquerade as being part of Shopify and not an app so they redirect to the marketing section.

Reply:
I thought the return_to would be the path the user gets taken to within the app iframe after the login flow finishes.

Reply:
Oh. Okay, I'll make the changes.